File: webview/terminal/src/render/snapshot.ts
import { base64ToBytes, writeBytes, type TerminalWriter } from "@/render/delta";
import type {
  AnchorState,
  CharsetState,
  Color,
  MouseMode,
  SgrAttrs,
  Snapshot,
} from "@/types/frame";

interface SnapshotTerminal extends TerminalWriter {
  reset(): void;
  resize(cols: number, rows: number): void;
}

const ESC = "\x1b";
const BEL = "\x07";
const SI = "\x0f";
const SO = "\x0e";

export async function applySnapshot(
  term: SnapshotTerminal,
  snap: Snapshot,
): Promise<void> {
  term.reset();
  term.resize(snap.cols, snap.rows);

  const anchorBytes = encodeAnchorState(snap.anchor_state);
  if (anchorBytes.length > 0) {
    await writeBytes(term, anchorBytes);
  }

  await writeBytes(term, base64ToBytes(snap.bytes_b64));
}

function encodeAnchorState(anchor: AnchorState): Uint8Array {
  return new TextEncoder().encode(
    [
      encodeTitle(anchor.title),
      anchor.modes.deccmm_cursor_keys ? encodeMode("1", true) : "",
      anchor.modes.autowrap ? "" : encodeMode("7", false),
      anchor.modes.alt_screen ? encodeMode("1049", true) : "",
      anchor.modes.bracketed_paste ? encodeMode("2004", true) : "",
      encodeMouseMode(anchor.modes.mouse_reporting),
      anchor.modes.origin_mode ? encodeMode("6", true) : "",
      encodeSgr(anchor.sgr),
      encodeCharset(anchor.charset),
      encodeCursor(anchor.cursor),
    ].join(""),
  );
}

function encodeTitle(title: string | null): string {
  return title === null ? "" : `${ESC}]0;${sanitizeTitle(title)}${BEL}`;
}

function sanitizeTitle(title: string): string {
  return Array.from(title)
    .filter((char) => {
      const codePoint = char.codePointAt(0);
      return (
        codePoint !== undefined &&
        !(
          (codePoint >= 0x00 && codePoint <= 0x1f) ||
          (codePoint >= 0x7f && codePoint <= 0x9f)
        )
      );
    })
    .join("");
}

function encodeMode(mode: string, enabled: boolean): string {
  return `${ESC}[?${mode}${enabled ? "h" : "l"}`;
}

function encodeMouseMode(mouseMode: MouseMode): string {
  switch (mouseMode) {
    case "Off":
      return "";
    case "X10":
      return encodeMode("9", true);
    case "Normal":
      return encodeMode("1000", true);
    case "ButtonEvent":
      return encodeMode("1002", true);
    case "AnyEvent":
      return encodeMode("1003", true);
  }
}

function encodeSgr(sgr: SgrAttrs): string {
  const codes = [
    sgr.bold ? "1" : null,
    sgr.faint ? "2" : null,
    sgr.italic ? "3" : null,
    sgr.underline ? "4" : null,
    sgr.blink ? "5" : null,
    sgr.reverse ? "7" : null,
    sgr.strikethrough ? "9" : null,
    ...encodeColor(sgr.fg, false),
    ...encodeColor(sgr.bg, true),
  ].filter((code): code is string => code !== null);

  return `${ESC}[0m${codes.length > 0 ? `${ESC}[${codes.join(";")}m` : ""}`;
}

function encodeColor(color: Color | null, background: boolean): string[] {
  if (color === null) {
    return [];
  }

  if ("Palette" in color) {
    if (color.Palette >= 0 && color.Palette <= 7) {
      return [String(color.Palette + (background ? 40 : 30))];
    }

    if (color.Palette >= 8 && color.Palette <= 15) {
      return [String(color.Palette - 8 + (background ? 100 : 90))];
    }

    return [background ? "48" : "38", "5", String(color.Palette)];
  }

  if ("Indexed" in color) {
    return [background ? "48" : "38", "5", String(color.Indexed)];
  }

  return [
    background ? "48" : "38",
    "2",
    String(color.Rgb[0]),
    String(color.Rgb[1]),
    String(color.Rgb[2]),
  ];
}

function encodeCharset(charset: CharsetState): string {
  const designators = ["(", ")", "*", "+"];
  const designations = charset.g
    .map((finalByte, index) => `${ESC}${designators[index]}${String.fromCharCode(finalByte)}`)
    .join("");

  // xterm exposes SI/SO for GL G0/G1. Restoring arbitrary GL G2/G3 and GR
  // safely would require invasive terminal state changes, so leave them intact.
  const glShift = charset.gl === 0 ? SI : charset.gl === 1 ? SO : "";

  return `${designations}${glShift}`;
}

function encodeCursor(cursor: [number, number]): string {
  const [col, row] = cursor;
  return `${ESC}[${row + 1};${col + 1}H`;
}

File: crates/client/client-core/src/snapshot.rs
use std::fmt::Write as _;

use bytes::Bytes;
use cli_pocket_proto::{
    AnchorState, CharsetState, Color, MouseMode, SgrAttrs, StreamSeq, TerminalInfo, TerminalModes,
};

#[derive(Debug, Clone)]
pub struct TerminalSnapshot {
    pub info: TerminalInfo,
    pub start_seq: StreamSeq,
    pub end_seq: StreamSeq,
    pub bytes: Bytes,
    pub render_prefix: String,
}

impl TerminalSnapshot {
    #[must_use]
    pub fn new(
        info: TerminalInfo,
        start_seq: StreamSeq,
        end_seq: StreamSeq,
        bytes: Bytes,
        render_prefix: String,
    ) -> Self {
        Self {
            info,
            start_seq,
            end_seq,
            bytes,
            render_prefix,
        }
    }
}

#[must_use]
pub fn render_prefix_from_anchor(anchor: &AnchorState) -> String {
    let mut out = String::new();
    out.push_str("\x1b[0m");
    out.push_str("\x1b[?25h");

    apply_sgr(&mut out, &anchor.sgr);
    apply_modes(&mut out, anchor.modes);
    apply_charset(&mut out, anchor.charset);

    let row = anchor.cursor.1.saturating_add(1);
    let col = anchor.cursor.0.saturating_add(1);
    let _ = write!(out, "\x1b[{row};{col}H");

    out
}

fn apply_sgr(out: &mut String, sgr: &SgrAttrs) {
    let mut params = Vec::new();
    if sgr.bold {
        params.push("1".to_owned());
    }
    if sgr.faint {
        params.push("2".to_owned());
    }
    if sgr.italic {
        params.push("3".to_owned());
    }
    if sgr.underline {
        params.push("4".to_owned());
    }
    if sgr.blink {
        params.push("5".to_owned());
    }
    if sgr.reverse {
        params.push("7".to_owned());
    }
    if sgr.strikethrough {
        params.push("9".to_owned());
    }
    if let Some(fg) = sgr.fg {
        params.extend(color_params(fg, true));
    }
    if let Some(bg) = sgr.bg {
        params.extend(color_params(bg, false));
    }

    if !params.is_empty() {
        out.push_str("\x1b[");
        out.push_str(&params.join(";"));
        out.push('m');
    }
}

fn color_params(color: Color, foreground: bool) -> Vec<String> {
    match color {
        Color::Palette(index) => {
            let base = if foreground {
                if index < 8 {
                    30
                } else {
                    90
                }
            } else if index < 8 {
                40
            } else {
                100
            };
            vec![(base + i32::from(index % 8)).to_string()]
        }
        Color::Indexed(index) => vec![
            if foreground { "38" } else { "48" }.to_owned(),
            "5".to_owned(),
            index.to_string(),
        ],
        Color::Rgb(r, g, b) => vec![
            if foreground { "38" } else { "48" }.to_owned(),
            "2".to_owned(),
            r.to_string(),
            g.to_string(),
            b.to_string(),
        ],
    }
}

fn apply_modes(out: &mut String, modes: TerminalModes) {
    set_private_mode(out, 1, modes.deccmm_cursor_keys);
    set_private_mode(out, 6, modes.origin_mode);
    set_private_mode(out, 7, modes.autowrap);
    set_private_mode(out, 1000, !matches!(modes.mouse_reporting, MouseMode::Off));
    set_private_mode(
        out,
        1002,
        matches!(modes.mouse_reporting, MouseMode::ButtonEvent),
    );
    set_private_mode(
        out,
        1003,
        matches!(modes.mouse_reporting, MouseMode::AnyEvent),
    );
    set_private_mode(out, 1005, false);
    set_private_mode(out, 1006, !matches!(modes.mouse_reporting, MouseMode::Off));
    set_private_mode(out, 1049, modes.alt_screen);
    set_private_mode(out, 2004, modes.bracketed_paste);
}

fn set_private_mode(out: &mut String, mode: u16, enabled: bool) {
    out.push_str("\x1b[?");
    out.push_str(&mode.to_string());
    out.push(if enabled { 'h' } else { 'l' });
}

fn apply_charset(out: &mut String, charset: CharsetState) {
    for (index, set) in charset.g.iter().copied().enumerate() {
        let selector = match index {
            0 => '(',
            1 => ')',
            2 => '*',
            3 => '+',
            _ => continue,
        };
        out.push('\x1b');
        out.push(selector);
        out.push(char::from(set));
    }

    if charset.gl <= 3 {
        out.push('\x0f');
        if charset.gl != 0 {
            out.push('\x1b');
            out.push('n');
        }
    }
}
